org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Scanner.java

```diff
@@ -2146,20 +2000,39 @@ protected int scanForStringLiteral() throws InvalidInputException {
 			}
 			throw e; // rethrow
 		}
+		if (this.stringTemplateComponents != null) {
+			// mark the ending of the last fragment in the string template (before the closing quote)
+            if (this.jumpingOverEmbeddedExpression == 0) {
+			    int textFragmentEnd = this.currentPosition - 2;
+			    addStringTemplateComponent(new TextFragment(textFragmentStart, this.currentPosition - 1 , getCurrentTokenInRange(textFragmentStart - this.startPosition, textFragmentEnd - this.startPosition)));
+            }
+			return TokenNameStringTemplate;
+		}
 		return TokenNameStringLiteral;
 	}
 }
 
 protected int scanForTextBlock() throws InvalidInputException {
 	int lastQuotePos = 0;
 	try {
-		this.rawStart = this.currentPosition - this.startPosition;
+		this.textBlockOffset = this.currentPosition - this.startPosition;
+		int textFragmentStart = this.startPosition;
+		int lastScannedPos = this.currentPosition;
 		while (this.currentPosition <= this.eofPosition) {
 			if (this.currentCharacter == '"') {
 				lastQuotePos = this.currentPosition;
 				// look for text block delimiter
 				if (scanForTextBlockClose()) {
-					this.currentPosition += 2;
+					if (this.stringTemplateComponents != null) {
+						// mark the ending of the last fragment in the string template (before the closing quote)
+			            if (this.jumpingOverEmbeddedExpression == 0) {
+						    addStringTemplateComponent(new TextFragment(textFragmentStart, this.currentPosition + 1,
+						    		getCurrentTokenInRange(textFragmentStart - this.startPosition, this.currentPosition - this.startPosition - 2)));
+			            }
+			            this.currentPosition += 2; // move it past the triple quote
+						return TokenNameTextBlockTemplate;
```

Comment:
Doubtful if this will work for unicode usage for """

Reply:
OK, I just checked with Javac and it appears quotes as unicode are not valid delimiters for text blocks. So, I think we are safe here.